core/src/main/java/com/linecorp/armeria/common/HttpMessageAggregator.java

```diff
@@ -109,14 +119,32 @@ public void accept(Void unused, Throwable cause) {
         if (contentLength == 0) {
             content = HttpData.EMPTY_DATA;
         } else {
-            final byte[] merged = new byte[contentLength];
-            for (int i = 0, offset = 0; i < contentList.size(); i++) {
-                final HttpData data = contentList.set(i, null);
-                final int dataLength = data.length();
-                System.arraycopy(data.array(), data.offset(), merged, offset, dataLength);
-                offset += dataLength;
+            if (alloc != null) {
+                final ByteBuf merged = alloc.buffer(contentLength);
+                for (int i = 0; i < contentList.size(); i++) {
+                    final HttpData data = contentList.set(i, null);
+                    if (data instanceof ByteBufHolder) {
+                        ByteBufHolder byteBufData = (ByteBufHolder) data;
+                        try {
+                            merged.writeBytes(byteBufData.content());
```

Comment:
Just curious. Is [CompositeByteBuf](https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java) not suitable for this case?

Reply:
I have gone back and forth on usage of `CompositeByteBuf` - I feel like it's great when the number of buffers is unbounded since the number of copies is unknown, but when we know the length, trading off exactly one copy to avoid the overhead of `CompositeByteBuf` (ArrayList, managing indexes over buffer boundaries) is probably often worth it. Perfect might be adding a heuristic that uses it based on the number of elements / contentLength here, but getting quite deep at that point.